include/SFML/Window/WindowBase.inl

```diff
@@ -0,0 +1,73 @@
+////////////////////////////////////////////////////////////
+//
+// SFML - Simple and Fast Multimedia Library
+// Copyright (C) 2007-2024 Laurent Gomila (laurent@sfml-dev.org)
+//
+// This software is provided 'as-is', without any express or implied warranty.
+// In no event will the authors be held liable for any damages arising from the use of this software.
+//
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it freely,
+// subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented;
+//    you must not claim that you wrote the original software.
+//    If you use this software in a product, an acknowledgment
+//    in the product documentation would be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such,
+//    and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution.
+//
+////////////////////////////////////////////////////////////
+
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#include <SFML/Window/Event.hpp>
+#include <SFML/Window/WindowBase.hpp> // NOLINT(misc-header-include-cycle)
+
+#include <variant>
+
+
+namespace sf
+{
+namespace priv
+{
+template <typename... Ts>
+struct OverloadSet : Ts...
+{
+    using Ts::operator()...;
+};
+template <typename... Ts>
+OverloadSet(Ts...) -> OverloadSet<Ts...>;
+
+template <typename... Ts>
+struct OverloadSetWithDefault : OverloadSet<Ts...>
+{
+    // By providing our own operator() and forwarding based
+    // on invocability of OverloadSet<Ts...> on the concrete type
+    // of the value, we save the user from having to provide
+    // their own catch-all overload if they don't want to
+    template <typename T>
+    void operator()([[maybe_unused]] T&& value) // NOLINT(cppcoreguidelines-missing-std-forward)
+    {
+        if constexpr (std::is_invocable_v<OverloadSet<Ts...>, T>)
+            OverloadSet<Ts...>::operator()(std::forward<T>(value));
+    }
+};
+template <typename... Ts>
+OverloadSetWithDefault(Ts...) -> OverloadSetWithDefault<Ts...>;
+} // namespace priv
+
+
+////////////////////////////////////////////////////////////
+template <typename... Ts>
+void WindowBase::handleEvents(Ts&&... handlers) // NOLINT(cppcoreguidelines-missing-std-forward)
+{
+    while (const std::optional event = pollEvent())
```

Comment:
What if the user wants event visitation and `waitEvent`?
Couldn't `visit` simply be a method of the `sf::Event` class?

Reply:
Making visit part of `sf::Event` would mean the user would still have to pick up the events themselves and check for validity. `handleEvents` handles pending events without waiting by design. The main goal of this addition is to alleviate the majority of users from having to deal with validity checking completely. The users who need to wait for events in their application can still handle them the "classical" way using `waitEvent`. I consider using `waitEvent` "advanced usage" so it is also acceptable for those who need that functionality to have to implement more of the event handling themselves.

Reply:
> What if the user wants event visitation and waitEvent?
Couldn't visit simply be a method of the sf::Event class?

That's a good point. It begs the question why we don't just handle visitation through the `sf::Event` API like previous proposals. It would let us avoid using a `friend` which obscure the relationship between classes.

```cpp
while (const std::optional event = window.{pollEvent or waitEvent}())
    event->visit([&](const sf::Event::Closed&){ ... }, ...);
```

> I consider using waitEvent "advanced usage"

I consider visitation to be "advanced usage" as well. It takes a certain level of C++ proficiency to structure your app around callbacks rather than more conventional control flow structures.

Reply:
> I consider visitation to be "advanced usage" as well. It takes a certain level of C++ proficiency to structure your app around callbacks rather than more conventional control flow structures.

Which is why this change goes out of its way to keep this concept under the hood. For a beginning user, they will not have to know what a variant is, nor what visitation means. The fact that _we_ know what is happening behind the scenes doesn't make it harder for beginners to learn. You can pick up any beginner Java book and this is a pattern they will learn on day 1 when designing UI applications. The concept of passing callables to a function is not a hard one to learn, but as usual when some straight-forward concept is introduced to C++-land it becomes "spooky" and for whatever reason we start to worry that it will be too hard for beginners to grasp.

Having to check less things in order to get code that works will always be more beginner friendly in my book. If anyone has read "Programming: Principles and Practice Using C++", they will know that dynamic allocation only comes deep into the latter half of the book, for good reason. Even though we know that a `std::optional` doesn't dynamically allocate, the fact that you have to check if it is valid and then use the same operator to "dereference" it as you would a pointer just screams "advanced concept" to me.

I find it worrying that there has been a trend lately that disregards the little things like this. While a beginning user was perfectly capable of writing complete games in SFML 2 without having to use the `->` operator a single time in their code base or having to check if any "pointer-like-object" was valid, we just litter the current API at every corner we can with these concepts all in the name of "safety". What use is a safe API if the users we are targeting don't even know what half the symbols on the screen mean? I would even go so far as to claim that the new API would lead to even more "problems" if a novice user who is unfamiliar with these concepts were to jump head first into it. Sure we could claim that it is still "safer" but replacing one problem with another is still not going to leave a good impression with the user whose program doesn't work for reasons they don't yet understand.

I guess I'll have to repeat the motivation of this change once more: This change aims to reduce the amount of assumed knowledge required to handle events properly and also reduce the number of mandatory "boilerplate" steps required to write code that doesn't crash. If having to write less code to achieve the same goal makes the experience more challenging and error prone then I don't know what else to say.

If the main concern is that the `waitEvent` enjoyers are left in the rain with this change, we can still add another visitation function for them as well. Just don't make the beginner experience more horrible than it has to be because of some really niche use cases.

Reply:
`visit` has been added to `sf::Event` and is used by `handleEvents` in the current revision.